### ISSUE TITLE
MPS backend gradient correctness issues with large shapes

### ISSUE THREAD

**Issue by jhavukainen** [pytorch/pytorch#153957]
### 🐛 Describe the bug

Gradients populated in the backward pass differ significantly between MPS and CPU when the shapes of the relevant operations reach large enough sizes. Demonstrated in the example with two consecutive linear op layers where the gradient of the output layer matches the CPU but the inner layer differs significantly. Decreasing the shape decreases the error until at small shapes the problem is hidden as it goes below the assertion tolerance threshold.

```
import torch
import copy
from torch.testing import assert_close

torch.manual_seed(123)

#Shapes reproducing the mismatch
n_samples, context_len, emb_size = 2,256,130
vocab_size = 50257

#Example shapes to pass the test
#n_samples, context_len, emb_size = 1,4,8
#vocab_size = 6675

inp_shape = (n_samples, context_len, emb_size)

x_cpu = torch.randn(inp_shape, dtype=torch.float)
x_mps = x_cpu.detach().clone().to('mps')

model_cpu = torch.nn.Sequential(
    torch.nn.Linear(emb_size, emb_size, bias=False),
    torch.nn.Linear(emb_size, vocab_size, bias=False),
)

model_mps = copy.deepcopy(model_cpu).to('mps')

cpu_res = model_cpu(x_cpu)
mps_res = model_mps(x_mps)

cpu_targ = torch.randn_like(cpu_res)
mps_targ = cpu_targ.detach().clone().to('mps')

cpu_loss = (cpu_res-cpu_targ).sum()
mps_loss = (mps_res-mps_targ).sum()

torch.mps.synchronize()
assert_close(cpu_loss, mps_loss.to('cpu'))

cpu_loss.backward()
mps_loss.backward()

torch.mps.synchronize()

cpu_dict = {}
mps_dict = {}

for name, param in model_cpu.named_parameters():
    if param.requires_grad:
        cpu_dict[name] = param.grad

for name, param in model_mps.named_parameters():
    if param.requires_grad:
        mps_dict[name] = param.grad

for name in cpu_dict.keys():
    print(f"Name: {name}")
    close = assert_close(cpu_dict[name], mps_dict[name].to('cpu'))
```

### Versions

Collecting environment information...
PyTorch version: 2.8.0.dev20250520
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.5 (arm64)
GCC version: Could not collect
Clang version: 16.0.0 (clang-1600.3.2.7)
CMake version: version 4.0.0
Libc version: N/A

Python version: 3.11.8 | packaged by conda-forge | (main, Feb 16 2024, 20:49:36) [Clang 16.0.6 ] (64-bit runtime)
Python platform: macOS-15.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Max

**Comment by jhavukainen** [comment]
Issue reproduces at least since `torch==2.3.0` so it doesn't seem like a recent regression. Reproduces on both MacOS14 and MacOS15.

**Comment by jhavukainen** [comment]
Update on the issue and a slightly leaner repro case:

Currently it looks like the issue comes from `grad_output` tensor receive as an input to the `mps_linear_backward` being issued too small of a MTLBuffer to back it. Regardless of the sizes and shapes used the buffer assigned for it has been created for a scalar and has the smallest available size of 256 bytes. This means that eventually any values going over the 256 bytes may get corrupted and the issue shows up more consistently the larger the actual tensor is.

So far I've verified that when the buffer object is created the request for MPSAllocator is for a scalar so it correctly gets assigned the minimal buffer. Currently tracking down where the gradient tensors by the backward pass engine to figure out the reason for the scalar sized buffer.

```
import torch
import copy
from torch.testing import assert_close

torch.manual_seed(123)

#Shapes reproducing the mismatch
context_len, emb_size = 128,53 #between 52 and 53 error explodes
vocab_size = 50273

# Shows up clearly on fp16. But reproduces on FP32 as well
dtype=torch.float32

inp_shape = (context_len, emb_size)

x_cpu = torch.ones(inp_shape, dtype=dtype)
x_cpu.requires_grad_(True)

x_mps = x_cpu.detach().clone().to('mps')
x_mps.requires_grad_(True)

model_cpu = torch.nn.Sequential(
    torch.nn.Linear(emb_size, vocab_size, bias=False, dtype=dtype),
)

for param in model_cpu.parameters():
    torch.nn.init.constant_(param.data, 1.0)

model_mps = copy.deepcopy(model_cpu).to('mps')


cpu_res = model_cpu(x_cpu)
mps_res = model_mps(x_mps)

cpu_targ = torch.ones_like(cpu_res)
mps_targ = cpu_targ.detach().clone().to('mps')

cpu_loss = 0.001*(cpu_res-cpu_targ).sum()
mps_loss = 0.001*(mps_res-mps_targ).sum()
mps_loss.retain_grad()

torch.mps.synchronize()
assert_close(cpu_loss, mps_loss.to('cpu'))

cpu_loss.backward()
mps_loss.backward()

torch.mps.synchronize()

print("*"*50)
print("Mismatch in gradient at the first column")
print("CPU")
print(x_cpu.grad[:][0])
print("MPS")
print(x_mps.grad[:][0])
print("*"*50)

print("*"*50)
print("Other columns match. Here's the last column.")
print("CPU")
print(x_cpu.grad[:][-1])
print("MPS")
print(x_mps.grad[:][-1])
print("*"*50)
```
